### ISSUE TITLE
Notification audio not working

### ISSUE THREAD

**Issue by PlatinumSeed** [slack-for-linux/slack-for-linux#22]
Hi, I'm using ubuntu 14.10 and the audio notifications are not working. This includes previewing them under preferences.

**Comment by LuckyThirteen** [comment]
:+1:

**Comment by twolfson** [comment]
I can reproduce the error. Steps to reproduce:

1. Log into a valid slack team
2. Access account menu via Up arrow in lower-left corner next to username
3. Click on "Preferences" in account menu
4. Navigate to "Notifications" tab in modal
5. Click "Preview" under the Sounds heading

Expected behavior: A sound plays

Actual behavior: Nothing happens

For debugging, I recommend we check out the web client in Chrome to verify sound does play. If it does, then we might need to enable Flash. Keep on digging around on the page or JS source code to see what makes it tick.

Does anyone want to take this on? If not, I will take a look by the end of next weekend.

**Comment by twolfson** [comment]
I have decided to triage this sooner rather than later since the fix could be fast. It looks like we lack MP3 support. I wound up digging into the source code and finding `TS.sounds`/`soundManager`. The `soundManager` seems to be picking up HTML5 audio which is great.

```js
> soundManager.version
"V2.97a.20130512 (HTML5-only mode)"
```

Enabling Flash didn't fix the problem in case anything was lying. I wound up going to an HTML5 audio test page and found out that `mp3` isn't supported out of the box.

http://hpr.dogphilosophy.net/test/

![selection_107](https://cloud.githubusercontent.com/assets/902488/6767294/97609732-cff6-11e4-8458-c2ae33d0dd86.png)

Then, with a bit of research, I found out that we need to follow these steps:

https://github.com/nwjs/nw.js/wiki/Using-MP3-&-MP4-%28H.264%29-using-the--video--&--audio--tags.

**Comment by twolfson** [comment]
After reading up on it more, it looks like our grunt builder recommends using `grunt-contrib-copy` for the `.so`. However, by using the `.so` it sounds like we need to double license with GPL.


https://github.com/nwjs/nw.js/wiki/Using-MP3-&-MP4-%28H.264%29-using-the--video--&--audio--tags

Here are the details on GPL for anyone curious:

http://choosealicense.com/licenses/gpl-2.0/

**Comment by twolfson** [comment]
I am going to look into using Flash which will already have been licensed by the user and not put any legal weight on our repo.

**Comment by twolfson** [comment]
As a sanity check, I looked if there were other music formats. However, Slack only seems to have mp3's =/

![selection_108](https://cloud.githubusercontent.com/assets/902488/6767357/4b2fe036-cff9-11e4-99a4-ffb20f6c2ff2.png)

Fun fact: In the source code there is a `frog.mp3` for `beep`.

```js
      if (c == "beep") {
        c = "frog.mp3"
      }
```

**Comment by twolfson** [comment]
Alright, looks like Flash is possible. Slack is using SoundManager2:

http://www.schillmania.com/projects/soundmanager2/doc/#

To get a new sound playing in the Slack window, we had to re-setup the application with `preferFlash`, reboot it, and re-register the sound:

```js
soundManager.setup({preferFlash: true});
soundManager.reboot();
soundManager.createSound({url: 'https://slack.global.ssl.fastly.net/dfc0/sounds/push/flitterbug.mp3'}).play();
```

Slack caches sounds in a private variable though so I am not sure this is a viable option.

**Comment by twolfson** [comment]
It is viable if we are willing to overload `TS.sounds.play` with a function of the same signature.

http://viewsource.in/https://slack.global.ssl.fastly.net/31971/js/rollup-client_1420067921.js#L3972-3983

@wlaurance What is your preference: Use the `.so` and double license with GPL or overload the sound playback function?

**Comment by wlaurance** [comment]
I'd like to what is better for the long term maintainability and ease of implementing.

The GPL just means anyone has to keep this free an open source right?

**Comment by twolfson** [comment]
Yep, any forks have to be open source and state any significant changes:

http://choosealicense.com/licenses/gpl-2.0/

**Comment by twolfson** [comment]
I am pretty swamped on other projects at the moment. Not sure when I will get to this. If I do pick it up, I will post a comment before hand. Otherwise, someone feel to post a PR, the instructions are clear cut here:

https://github.com/mllrsohn/grunt-node-webkit-builder/wiki/How-to-automate-node-webkit-support-for-MP3-&-H264-in-video-&-audio-tags-with-grunt-contrib-copy

**Comment by wlaurance** [comment]
@PlatinumSeed Hey, thanks for all of your hard work on this project. I don't think it'd be much without you.

I've been meaning to get involved with this again. I've been swamped. I'm hoping to make a pass in the next two weeks.

**Comment by twolfson** [comment]
While debugging another issue, I noticed that the downloaded cache files have `libffmpegsumo.so` in them. I am guessing we are supposed to use that.

**Comment by twolfson** [comment]
```bash
$ ls cache/0.12.0/linux64/
credits.html      libffmpegsumo.so  nw                nw.pak            
icudtl.dat        locales/          nwjc     
```

**Comment by twolfson** [comment]
@wlaurance Based on your comments in #56, are you still planning to take on this bug or should someone else claim it?

**Comment by wlaurance** [comment]
Let's add a help wanted tag to this one.

**Comment by bronzdoc** [comment]
Hi, i've been working on this issue, i manage to fix it by replacing the wrong "libffmpegsumo.so" with the the correct one(got it from "chromium-codecs-ffmpeg-extra" package), i found two ways of fixing it , one way is: following this https://github.com/mllrsohn/grunt-node-webkit-builder/wiki/How-to-automate-node-webkit-support-for-MP3-&-H264-in-video-&-audio-tags-with-grunt-contrib-copy as you point out,
but i notice that you removed "grunt-node-webkit-builder" in previous commits and this solution requires to replace the .so after the build.

Another way i manage to fix it, is by replacing the .so in the "nw" directory in node modules, that will require to run a postinstall script that replaces the .so, let me know if you agree with this solution to send a pull request or if you had something else in mind, or if i'm missing something.

**Comment by twolfson** [comment]
Yep, the second solution sounds like the best option. For reference, the `grunt-node-webkit-builder` wiki links to `nw.js'` wiki page which has the same answer =)

https://github.com/nwjs/nw.js/wiki/Using-MP3-&-MP4-%28H.264%29-using-the--video--&--audio--tags.#linux

**Comment by twolfson** [comment]
This has been fixed by @bronzdoc in `1.6.0`. Please upgrade to the latest version to get audio notification support.

```bash
npm install npm@latest -g
```
